### PR TITLE
Adding the egs_lattice library with example geom

### PR DESCRIPTION
**Pull request by MartinMartinov**
Added new egs_lattice geometry useful for repeating a single geometry
indefinitely inside another geometry without storing each instance
of the repeated geometry separately.  Example uses:
```
:start geometry:
       library           = egs_lattice
       name              = phantom_w_microcavity
       base geometry     = phantom
       subgeometry       = microcavity
       subgeometry index = 0
       spacing           = 1 2 3
:stop geometry:

:start geometry:
       library           = egs_lattice
       name              = phantom_w_microcavity
       base geometry     = phantom
       subgeometry       = microcavity
       subgeometry index = 0
       spacing           = 1
:stop geometry:

:start geometry:
       library           = egs_lattice
       type              = hexagonal
       name              = phantom_w_microcavity
       base geometry     = phantom
       subgeometry       = microcavity
       subgeometry index = 0
       spacing           = 1
:stop geometry:
```
To be tested:
 - RelativeRho functions
 - BScaling functions
 - BooleanProperty functions
 - getLabelRegions

### CODE REVIEW

File: HEN_HOUSE/egs++/geometry/examples/example_egs_lattice.geom [hunk 1]
```diff
@@ -0,0 +1,113 @@
+/*
+###############################################################################
+#
+#  EGSnrc egs++ lattice geometry
+#  Copyright (C) 2019 Rowan Thomson and Martin Martinov
+#
+#  This file is part of EGSnrc.
+#
+#  egs_lattice is free software: you can redistribute it and/or modify it under
```
Comment (ftessier) [inline HEN_HOUSE/egs++/geometry/examples/example_egs_lattice.geom]: Oops, the licence text here should not be modified, i.e., should say "EGSnrc is free software..." (same thing in next paragraph). We don't want to be updating that text in every file, so we chose to write: "This file is part of EGSnrc. EGSnrc is free software...".

**Comment by rtownson** [comment]
The egs_lattice geometry needs to be added to the geometries list in the egs++ `Makefile`.

File: HEN_HOUSE/egs++/geometry/egs_lattice/egs_lattice.h [hunk 2]
```diff
@@ -0,0 +1,898 @@
...
+#
+#  Contributors:
+#
+###############################################################################
+*/
+
+
+/*! \file egs_lattice.cpp
+ *  \brief A Bravais, cubic, and hexagonal lattice geometry
+
```
Comment (rtownson) [inline HEN_HOUSE/egs++/geometry/egs_lattice/egs_lattice.h:39]: We should have `\ingroup Geometry` here. Also a general description of the geometry, and each of the inputs.

Comment (rtownson) [reply HEN_HOUSE/egs++/geometry/egs_lattice/egs_lattice.h:41]: Also `\ingroup CompositeG`

File: HEN_HOUSE/egs++/geometry/egs_lattice/egs_lattice.h [hunk 3]
```diff
@@ -0,0 +1,920 @@
...
+into one region of a second geometry (named base geometry).  This geometry
+effectively recurses the subgeometry at every position defined by a Bravais,
+cubic, or hexagonal lattice.  As such, you can model an infinite amount of
+subgeometries (e.g., region 0 of egs_space) and the only slow down to your
+simulation would depend on how many subgeometries you would expect over a
+particle track.
+As this geometry only stores a single subgeometry in memory, it can only
+score in ALL subgeometries (or rather, the single subgeometry at all lattice
+positions) at once.  Therefore, dose to the subgeometry at different
+locations cannot be discerned.
```
Comment (rtownson) [inline HEN_HOUSE/egs++/geometry/egs_lattice/egs_lattice.h:54]: I would also add the paper reference here. Users a likely to be confused as to how this geometry differs from a simple repeater geometry.

**Comment by ftessier** [comment]
@MartinMartinov I am looking to merge this into the Release 2021 (or else into `develop` just after the release). Are you ok if I push to your `develop` branch to sync it up and get it ready to merge (couple tiny conflicts, formatting, etc.)?

**Comment by MartinMartinov** [comment]
That's absolutely fine with me.

**Review by mainegra — APPROVED** [review]

**Review by blakewalters — APPROVED** [review]

**Comment by ftessier** [comment]
I am not able to push changes to the @MartinMartinov branch. I will thus close this pull request and reopen a new one to push changes and merge.

**Review by ftessier — APPROVED** [review]

**Comment by ftessier** [comment]
Merged as #683.

**Pull request closed without merging**